### PR TITLE
Task/raja 307 scan functional tests

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR reworks RAJA scan tests and puts them in the new test organization.

### CODE REVIEW

**Comment by willkill07** [comment]
@rhornung67 Could I write the empty-range tests on this branch instead of #769 ?

**Comment by rhornung67** [comment]
@willkill07 absolutely.  We're trying to avoid rework as much as possible as we convert our tests to the new organization and increase coverage.  Thanks!

File: test/functional/scan/test-scan-exclusive-seq.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "test-scan-exclusive.hpp"
+
+// Sequential policy types to test
+using SequentialExecTypes = list< RAJA::seq_exec, 
+                                  RAJA::loop_exec >;
```
Comment (mdavis36) [inline test/functional/scan/test-scan-exclusive-seq.cpp]: Do we want RAJA::simd_exec here as well?

File: test/functional/scan/test-scan-exclusive-tbb.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "test-scan-exclusive.hpp"
+
+#if defined(RAJA_ENABLE_TBB)
+
+// TBB policy types to test
+using TBBExecTypes = list< RAJA::tbb_for_exec >;
```
Comment (mdavis36) [inline test/functional/scan/test-scan-exclusive-tbb.cpp]: RAJA::tbb_for_static and RAJA::tbb_for_dynamic?

File: test/functional/scan/test-scan-inclusive-seq.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "test-scan-inclusive.hpp"
+
+// Sequential policy types to test
+using SequentialExecTypes = list< RAJA::seq_exec, 
+                                  RAJA::loop_exec >;
```
Comment (mdavis36) [inline test/functional/scan/test-scan-inclusive-seq.cpp]: Same thing here, simd_exec?

File: test/functional/scan/test-scan-inclusive-tbb.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "test-scan-inclusive.hpp"
+
+#if defined(RAJA_ENABLE_TBB)
+
+// TBB policy types to test
+using TBBExecTypes = list< RAJA::tbb_for_exec >;
```
Comment (mdavis36) [inline test/functional/scan/test-scan-inclusive-tbb.cpp]: Same here tbb_for_static, tbb_for_dynamic

File: test/functional/scan/test-scan-exclusive-tbb.cpp (hunk 2 above)
Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: Thanks. I'll add that.

File: test/functional/scan/test-scan-inclusive-seq.cpp (hunk 3 above)
Comment (rhornung67) [reply test/functional/scan/test-scan-inclusive-seq.cpp]: Scans shouldn't work with simd (at least not yet).

File: test/functional/scan/test-scan-inclusive-tbb.cpp (hunk 4 above)
Comment (rhornung67) [reply test/functional/scan/test-scan-inclusive-tbb.cpp]: Got it. Thanks.

**Comment by rhornung67** [comment]
@willkill07 I can't figure out why parallel scans (OpenMP) do not work with plus operator and float data. This case was omitted from the original scan tests you wrote.  Any ideas about this?

**Comment by willkill07** [comment]
@rhornung67 -- as an educated guess the precision of float is too low given the input data. We cannot accurately compute the sum with 23 mantissa bits which then get reduced via OpenMP. This is one reason why I didn't test `float`.

In general terms, `((a + b) + c) != (a + (b + c))` when `a`, `b`, and `c` are `float`. Associativity is not guaranteed.

**Comment by rhornung67** [comment]
@willkill07 thanks for the explanation. We have the same issue with CUDA.  Do you think it's best to remove all scan tests with float data and just do ints and doubles?

File: test/functional/scan/test-scan-exclusive-tbb.cpp (hunk 2 above)
Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: @mdavis36 actually, tbb_for_static is not exposed as a user policy. tbb_for_exec is an alias to the TBB static policy.

Comment (mdavis36) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: I thought that tbb_for_exec just called the default of tbb_for_static where Chunksize = 1 and that tbb_for_static was for using a user defined Chunksize.

Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: I thought so too. But, when I added it, the code didn't compile. There is no use of tbb_for_static in or for all unit tests.  Did you get it to work for the forall tests? I will skip it for now and add an issue that we can look into.

Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: Wait. I think I was using it wrong.  Let me try again.

Comment (mdavis36) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: I did, it wont compile without the template argument, I am launching it as tbb_for_static<8>

**Comment by rhornung67** [comment]
@mdavis36 @rchen20 @davidbeckingsale @artv3 this PR should be good to go as soon as checks pass.  Please review when you get a chance.

@willkill07 should we wait to merge this until you add the empty range tests to this branch. Or do you want to do it on a separate PR?

File: test/functional/scan/test-scan-exclusive-tbb.cpp (hunk 2 above)
Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: Yup. I screwed up. I added a few tests for tbb_for_static.

File: test/functional/scan/test-scan-exclusive-tbb.cpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "test-scan-exclusive.hpp"
+
+#if defined(RAJA_ENABLE_TBB)
+
+// TBB policy types to test
+using TBBExecTypes = list< RAJA::tbb_for_exec,
+                           RAJA::tbb_for_static< >,
```
Comment (mdavis36) [inline test/functional/scan/test-scan-exclusive-tbb.cpp]: We might not need this one since tbb_for_exec aliases tbb_for_static<>

Comment (rhornung67) [reply test/functional/scan/test-scan-exclusive-tbb.cpp]: I thought about that. Then, I thought maybe in the future that may not be the case so I added it. It's not horribly redundant.

**Comment by rhornung67** [comment]
@mdavis36 @davidbeckingsale @ajkunen @rchen20 @artv3 @MrBurmark When you have a chance, could I please get some reviews on this PR?  Thanks.

File: host-configs/lc-builds/hip.cmake [hunk 6]
```diff
@@ -29,17 +29,24 @@ if (ENABLE_OPENMP)
 	set(HIP_COMMON_OPT_FLAGS "-fopenmp ${HIP_COMMON_OPT_FLAGS}")
 endif()
 
+#if(CMAKE_BUILD_TYPE MATCHES Release)
+#  set(RAJA_HIPCC_FLAGS -O2; ${HIP_COMMON_OPT_FLAGS}; ${HOST_OPT_FLAGS} CACHE LIST "")
+#elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
+#  set(RAJA_HIPCC_FLAGS -g; -G; -O2; ${HIP_COMMON_OPT_FLAGS}; ${HOST_OPT_FLAGS} CACHE LIST "")
+#elseif(CMAKE_BUILD_TYPE MATCHES Debug)
+#  set(RAJA_HIPCC_FLAGS -g; -G; -O0; ${HIP_COMMON_DEBUG_FLAGS}; CACHE LIST "")
+#endif()
```
Comment (MrBurmark) [inline host-configs/lc-builds/hip.cmake]: Should this be left here?

Comment (rhornung67) [reply host-configs/lc-builds/hip.cmake]: Good catch. I forgot to clean that up.  Thanks.

File: test/functional/scan/test-scan.hpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+// Scan functional test class
+template<typename T>
+class ScanFunctionalTest: public ::testing::Test {};
+
+
+// Define scan operation types
+using OpTypes = list< RAJA::operators::plus<int>,
+#if 0  // Parallel tests with plus operator and float data do not work
+       // likely due to precision being to low
```
Comment (MrBurmark) [inline test/functional/scan/test-scan.hpp]: typo

File: host-configs/lc-builds/hip.cmake (hunk 6 above)
Comment (rhornung67) [reply host-configs/lc-builds/hip.cmake]: Unfortunately, this went into develop yesterday. I just put up a PR to address that. After that is merged, that will propagate to this branch.

File: test/functional/scan/test-scan.hpp (hunk 7 above)
Comment (rhornung67) [reply test/functional/scan/test-scan.hpp]: to -> too fixed.  See what I get for adding comments.... :-)

**Comment by trws** [comment]
> @willkill07 thanks for the explanation. We have the same issue with CUDA. Do you think it's best to remove all scan tests with float data and just do ints and doubles?

I realize you weren't asking me, but I'd say do the int and double.  If we really need the float, we need to widen the error margin appropriately to allow for the non-determinism.

**Comment by willkill07** [comment]
I would advise having tests for all integral types and double. If doing single-precision point, we must assume the sum of all values can be represented within the fractional bits.  Otherwise we’d need to use some reasonable epsilon.

For example, values of uniform distribution of values `+- 2^i | abs(i)` and `N <= 2^(23-2*i)` will never overflow

I’ll be able to work on a PR for this to get merged into test reorg

**Comment by rhornung67** [comment]
@trws I understand what you are saying, but I will leave the min, max scan tests with float data in since they are working.  The weird thing (at least I think it's weird) is that the input arrays for the scan tests are initialized to integral values regardless of whether the array data type is float or double, yet the the prefix sum scans for float data failed with values that were off by a value of 2 IIRC. That is outside of a tolerance I would expect.

**Comment by willkill07** [comment]
@rhornung67 I will add the zero-length tests to this branch.

**Comment by rhornung67** [comment]
Thanks @willkill07.  Feel free to work in tests for zero-length ranges to that PR if you want.

**Comment by rhornung67** [comment]
@willkill07 you read my mind. 🥇

**Comment by rhornung67** [comment]
@mdavis36, @MrBurmark, @artv3, @rchen20, @ajkunen, @trws anyone else want to review this before I merge? We've got a few more PRs in the works to go into the test-reorg branch before we put up a PR to merge that into develop. They should be ready soon. We're trying to avoid dealing with folks adding too many new tests to the old-tests and then having to rework them into the new structure.  Thank you.
